### PR TITLE
Fixed regexps in show rules on text (via merging at evaluation)

### PR DESCRIPTION
**Pull request by sitandr**
Does the same as #1868, but merging is delayed until evaluation. The idea is generally the same. That allows using _string_ variables/functions inside text that show rule will match (see tests).

Unfortunately, there is another serious design problem: spans. Spans can only refer to existing single parsed nodes, so it is impossible to reference merged correctly. Editing nodes at evaluation step is impossible as evaluation works by immutable reference.

Currently I'm using span of _last_ node from merged text, but I guess that could lead to mess in some cases. I'm not properly aware of `Span` system, I guess.

Except that, everything seems good.

### CODE REVIEW

File: crates/typst/src/eval/mod.rs [hunk 1]
```diff
@@ -399,9 +399,81 @@ fn eval_markup<'a>(
...
+    let mut previous_text: Option<EcoString> = None;
+
+    // Let's believe that two adjacent spaces is gibberish.
+    // That variable helps to join two texts, if they are separated by spaces only
+    // And move space to it's own node otherwise.
+    let mut stored_space: Option<ast::Expr<'_>> = None;
+
+    // Unfortunately, it is currently impossible to attach exact span to merged text node,
+    // So let's attach at least some of them.
+    let mut last_span: Span = Span::detached();
```
Comment (PgBiel) [inline crates/typst/src/eval/mod.rs:413]: I hope my question isn't too ignorant, but is it possible to perhaps add joining functionality to Span to make this possible? I'm sure people would see it as a bug if Typst only highlighted up to the first escape of some error text or something (although that wouldn't be the end of the world, but worth considering).

Comment (sitandr) [reply crates/typst/src/eval/mod.rs:413]: Again, it's a design problem. Spans store their id-s that are created during parsing syntax nodes. Spans of children should be less than span of parent, and spans of next children should be larger than span of previous, something like that. Each expression stores &SyntaxNode that has it's own span.

I'm not sure, of course, but it seems to me that creating new spans dynamically during evaluation just isn't possible without changing how it works.

**Comment by laurmaedje** [comment]
I am not too worried about error spans since errors on text should be very rare. However, this will also break click-to-jump from preview to source. I did not think of this when proposing the alternative of doing this during evaluation.

In general, this is slightly different from what I had in mind. I though we would do this during addition / joining of content (`impl Add for Content`). This would generalized it even further. However, before thinking about whether that would be a better alternative, we need to consider what to do about the click-to-jump problem.

**Comment by sitandr** [comment]
Why will the jump `preview -> source` be broken?  After merge, it will be one big text node on preview, so after click you will move to the source of last subnode. I suppose it is fine.

But the jump `source -> preview` will be obviously broken. I'm not sure that jump to "edited place" is it, but if it is, it will jump only if you are editing the last text subnode. So if you start _adding_ something to the text, it will jump, but if you edit insides, it will not. Probably that could be even beneficial, like you don't want preview to jump if you just fix some typo. In general, when you editing some plain text, you don't need preview that much. But if you want it to, you can write something at paragraph end or just add some non-plain-text things, like bold.

**Comment by laurmaedje** [comment]
So in comparison:

1. In the parser approach, the jump is not really problematic. (For escape sequences, it could be a bit problematic). Text show rules can't match on text that isn't written literally and consecutively or with escape sequences.

2. In the evaluation approach, text show rules are more flexible, but source -> preview jump is very broken und preview -> source jump is somewhat broken. There remain cases where the match will not work.

3. In the realization approach, text show rules are more flexible and jumps would continue to work correctly (because text nodes wouldn't be merged / touched unless there is actually a match). Matching will pretty much always work. It's more complex to implement and one might consider it too powerful.

The parser approach is based on the mindset that text show rules operate on literal syntax whereas the later two approaches are based on the mindset that they operate on content, no matter where it stems from.

**Comment by sitandr** [comment]
Am I understanding that right, the third approach will not change the abstract syntax tree, so the problem of "randomly separated text nodes" will remain. Working with content with text from functions will be still much pain. But if we find match, then merge it, then pass the matched merged part to show rule, it will differ from that one will see when parsing that text "by hands". It can cause much confusion.

Also I'm afraid it can be quite inefficient. E.g., if we have five regexps, we will need to clone and join all adjacent text nodes for every one of them (or else we will get problems of the second approach).

Maybe I get it wrong, that approach isn't very clear to me.

**Comment by laurmaedje** [comment]
> If I understand that right, the third approach doesn't change the abstract syntax tree, so the problem of "randomly separated text nodes" will remain. Working with content with text from functions will be still much pain.

It is true that the randomly separated text nodes will remain from the perspective of a function that takes content. However, processing content in this way will always be difficult. A smart quote, for instance, would still split up text nodes even with your parser changes. So people doing this need to be prepared to handle the complications anyway. The sequence iteration approach is overall a bit flawed which is why content was originally opaque, but that wasn't great either.

> But if we find match, then merge it, then pass the matched merged part to show rule, it will differ from that one will see when parsing that text "by hands". It can cause much confusion.

Is this really that confusing? If you get a match, it's one text elements. Otherwise, it's content which can be anything. Parsing it "by hand" is typically just a bad idea.

> Also I'm afraid it can be quite inefficient. E.g., if we have five regexps, we will need to clone and join all adjacent text nodes for every one of them (or else we will get problems of the second approach).

The efficiency burdens can be overcome I believe. The approach I have in mind is like this:

1. We combine all active regexes into a single [regex automaton](https://docs.rs/regex-automata/latest/regex_automata/index.html). This way, we can efficiently search for all regexes at once. This step can be memoized so that each combination of active regexes is compiled just once. 

2. Regex-automaton DFAs support stream searching, so we can give it the text strings piece by piece or even byte by byte. There is no need for joining up or cloning strings. When there is a match, we need to backtrack and actually join up the strings. We'll probably need to do some bookkeeping to be able to join up the strings we already passed. Here we ruin the span a bit, but only for matches, so it should be okay.

**Comment by sitandr** [comment]
Makes sense, but sounds quite tricky. I'm really not sure if I will be able to do that in near time.

**Comment by laurmaedje** [comment]
I plan to refactor the realization soon-ish and might tackle it during this rework.

**Pull request closed without merging**